flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/FileStoreWriteImpl.java

```diff
@@ -82,18 +95,41 @@ public FileStoreWriteImpl(
     public RecordWriter createWriter(
             BinaryRowData partition, int bucket, ExecutorService compactExecutor) {
         Long latestSnapshotId = pathFactory.latestSnapshotId();
-        if (latestSnapshotId == null) {
-            return createEmptyWriter(partition, bucket, compactExecutor);
-        } else {
-            return createMergeTreeWriter(
-                    partition,
-                    bucket,
-                    scan.withSnapshot(latestSnapshotId)
-                            .withPartitionFilter(Collections.singletonList(partition))
-                            .withBucket(bucket).plan().files().stream()
-                            .map(ManifestEntry::file)
-                            .collect(Collectors.toList()),
-                    compactExecutor);
+        List<DataFileMeta> existingFileMetas = Lists.newArrayList();
+        if (latestSnapshotId != null) {
+            // Concat all the DataFileMeta of existing files into existingFileMetas.
+            scan.withSnapshot(latestSnapshotId)
+                    .withPartitionFilter(Collections.singletonList(partition)).withBucket(bucket)
+                    .plan().files().stream()
+                    .map(ManifestEntry::file)
+                    .forEach(existingFileMetas::add);
+        }
+
+        switch (writeMode) {
+            case APPEND_ONLY:
+                DataFilePathFactory factory =
+                        pathFactory.createDataFilePathFactory(partition, bucket);
+                FileStatsExtractor fileStatsExtractor =
+                        fileFormat.createStatsExtractor(writeSchema).orElse(null);
+
+                return new AppendOnlyWriter(
+                        fileFormat,
+                        options.targetFileSize,
+                        writeSchema,
+                        existingFileMetas,
+                        factory,
+                        fileStatsExtractor);
+
+            case CHANGE_LOG:
+                if (latestSnapshotId == null) {
+                    return createEmptyWriter(partition, bucket, compactExecutor);
+                } else {
+                    return createMergeTreeWriter(
+                            partition, bucket, existingFileMetas, compactExecutor);
+                }
```

Comment:
As we've calculated `existingFileMetas` we don't need a special `createEmptyWriter` method. Remove this method and this if branching.

Reply:
Make sense for me !